### DISCUSSION TITLE
Encode `typing.Literal` equality semantics (to catch common `x: str == MyLiteral` mistakes)

### DISCUSSION THREAD

**Discussion by thejcannon** [microsoft/pyright#7411] [Ideas]
Given this program:

```python
MyLiteral = Literal["hello!"]

def get_str() -> str:
   return "hello!"

if get_str() == MyLiteral:
    print("yay"). # unreachable code
```

According to the typesigs, yeah sure _anything_ can be compared and that'll return a boolean.

According to the semantics of `Literal`... we can probably assume that code is unreachable 😄 .

And I think this is a common (enough) mistake to warrant special-casing.

**Comment by erictraut** [comment]
Thanks for the suggestion. This is the first time we've received such a request, and I've never personally seen such an error, so I'm not (yet) convinced this is a "common mistake". It's something I'd consider adding in the future if we receive additional signal from pyright users (e.g. in the form of upvotes on this topic).
